File: Chapter 4/Methods Speed/objc-methods/Measure.m
//
//  Measure.m
//  Performance
//
//  Created by Kostiantyn Koval on 09/05/15.
//  Copyright (c) 2015 Kostiantyn Koval. All rights reserved.
//

#import "Measure.h"
@import QuartzCore;

@implementation Measure

+ (void)measure:(NSInteger)times call:(void(^)())call {
  NSMutableArray *results = [NSMutableArray new];

  for (NSInteger i = 0; i < times; i++) {
    CFTimeInterval startTime = CACurrentMediaTime();
    call();
    CFTimeInterval endTime = CACurrentMediaTime();
    [results addObject:@(endTime - startTime)];
  }
  CFTimeInterval avarage;
  for (NSNumber *time in results) {
    avarage += [time doubleValue];
  }
  avarage /= times;
  NSLog(@"Average Time - %f", avarage);
}

@end

File: Chapter 5/DataStructures/Speed Measurement/Helpers/Measure.swift
//
//  Logs.swift
//  Swift Toolkit
//
//  Created by Kostiantyn Koval on 02/05/15.
//  Copyright (c) 2015 Kostiantyn Koval. All rights reserved.
//

import Foundation
import QuartzCore

func measure(title: String? = nil, times: Int = 20, printOnlyAverage: Bool = true, call: () -> Void) {
  var results = [CFTimeInterval]()
  for _ in 1...times {
    let startTime = CACurrentMediaTime()
    call()
    let endTime = CACurrentMediaTime()
    let time =  endTime - startTime
    results.append(time)

    if !printOnlyAverage {
      print("Time - \(time)")
    }
  }

  if let title = title {
    print("\(title) : ", terminator: "")
  }
  let avarage = results.reduce(0, combine: +) / Double(times)
  print("Average - \(avarage)")
}
